lib/membrane/hls/source.ex

```diff
@@ -97,20 +99,56 @@ defmodule Membrane.HLS.Source do
     pad = Map.fetch!(state.ref_to_pad, ref)
     tracker = Map.fetch!(state.pad_to_tracker, pad)
 
-    case HLS.Storage.get_segment(state.storage, segment.uri) do
-      {:error, message} ->
-        Membrane.Logger.warn("HLS could not get segment #{inspect segment.uri}: #{inspect message}")
-        {{:ok, [{:redemand, pad}]}, state}
-        
-      {:ok, data} ->
-        action = {:buffer, {pad, %Buffer{payload: data, metadata: segment}}}
+    tracker =
+      tracker
+      |> Map.update!(:pending, &Q.push(&1, segment))
+      |> start_download(state.storage)
 
-        queue = Q.push(tracker.queue, action)
-        tracker = %{tracker | queue: queue}
-        state = %{state | pad_to_tracker: Map.put(state.pad_to_tracker, pad, tracker)}
+    state = %{state | pad_to_tracker: Map.put(state.pad_to_tracker, pad, tracker)}
+    {:ok, state}
+  end
 
-        {{:ok, [{:redemand, pad}]}, state}
-    end
+  def handle_other({task_ref, result}, _ctx, state) when is_reference(task_ref) do
+    # The task succeed so we can cancel the monitoring and discard the DOWN message
+    Process.demonitor(task_ref, [:flush])
+
+    {pad, tracker} = tracker_by_task_ref!(state.pad_to_tracker, task_ref)
+    segment = tracker.download.segment
+
+    tracker =
+      case result do
+        {:ok, data} ->
+          action = {:buffer, {pad, %Buffer{payload: data, metadata: segment}}}
+          ready = Q.push(tracker.ready, action)
+          %{tracker | ready: ready, download: nil}
+
+        {:error, message} ->
+          Membrane.Logger.warn(
+            "HLS could not get segment #{inspect(segment.uri)}: #{inspect(message)}"
+          )
+
+          %{tracker | download: nil}
+      end
+
+    tracker = start_download(tracker, state.storage)
```

Comment:
Is this call needed?

Reply:
The next download of the pending queue will start as soon as the segment was received.